### PR TITLE
Common StringBuilder

### PR DESCRIPTION
**Pull request by qurbonzoda**

### CODE REVIEW

**Comment by qurbonzoda** [comment]
The base branch for this PR will be changed to `master`.

File: runtime/src/main/kotlin/kotlin/text/StringBuilder.kt [hunk 1]
```diff
@@ -190,89 +304,340 @@ actual class StringBuilder private constructor (
...
+     * @param index the position in this string builder to insert at.
+     * @param csq the character sequence from which a subsequence is inserted. If [csq] is `null`,
+     *  then characters will be inserted as if [csq] contained the four characters `"null"`.
+     * @param start the beginning (inclusive) of the subsequence to insert.
+     * @param end the end (exclusive) of the subsequence to insert.
+     *
+     * @throws IndexOutOfBoundsException or [IllegalArgumentException] when [start] or [end] is out of range of the [csq] character sequence indices or when `start > end`.
+     * @throws IndexOutOfBoundsException if [index] is less than zero or greater than the length of this string builder.
+     */
+    @Deprecated("Renamed to `insertRange`", ReplaceWith("insertRange(indexm csq, start, end)"), DeprecationLevel.WARNING)
```
[the remark this answers is not in the document]
Comment (ilya-g) [reply runtime/src/main/kotlin/kotlin/text/StringBuilder.kt]: Also a typo i n replacement.

[the remark this answers is not in the document]
Comment (qurbonzoda) [reply runtime/src/main/kotlin/kotlin/text/StringBuilder.kt]: I'd rather remove the deprecation. @olonho, what do you think?

[the remark this answers is not in the document]
Comment (olonho) [reply runtime/src/main/kotlin/kotlin/text/StringBuilder.kt]: Is it perf tested?

[the remark this answers is not in the document]
Comment (qurbonzoda) [reply runtime/src/main/kotlin/kotlin/text/StringBuilder.kt]: No. There are no insert benchmarks for `StringBuilder`. Do you suppose the additional function invocation introduced tangible performance overhead?

[the remark this answers is not in the document]
Comment (olonho) [reply runtime/src/main/kotlin/kotlin/text/StringBuilder.kt]: StringBuilder is very performance sensitive, we need to ensure that there's no regressions here.

File: runtime/src/main/kotlin/kotlin/text/StringBuilder.kt [hunk 2]
```diff
@@ -142,44 +137,163 @@ actual class StringBuilder private constructor (
...
-    fun insert(index: Int, csq: CharSequence?): StringBuilder {
-        // Kotlin/JVM inserts the "null" string if the argument is null.
-        val toInsert = csq ?: "null"
-        return insert(index, toInsert, 0, toInsert.length)
+    /**
+     * Appends characters in the specified [chars] array to this string builder and returns this instance.
+     *
+     * Characters are appended in order, starting at the index 0.
+     */
+    actual fun append(chars: CharArray): StringBuilder {
```
[the remark this answers is not in the document]
Comment (qurbonzoda) [reply runtime/src/main/kotlin/kotlin/text/StringBuilder.kt]: It was decided to make those existing functions inline extension functions to not break user code

File: runtime/src/main/kotlin/kotlin/text/StringBuilder.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (qurbonzoda) [reply runtime/src/main/kotlin/kotlin/text/StringBuilder.kt]: I did benchmark StringBuilder methods, there are no regressions there.

**Comment by olonho** [comment]
Hm, but why all tests/builds fail?

**Comment by ilya-g** [comment]
@olonho It's intended to be merged into `kotlin-mirror/master` branch together with stdlib-common changes, so we're watching over this build: https://buildserver.labs.intellij.net/buildConfiguration/Kotlin_KotlinDev_SanityTestKotlinNativeLinuxComposite?branch=rr%2Fstdlib%2Fcommon-string-builder

However, it doesn't pass either. @qurbonzoda Could you rebase your changes onto the latest master and kotlin-mirror/master, where composite build is apparently green?

File: runtime/src/main/kotlin/kotlin/text/StringBuilder.kt [hunk 3]
```diff
@@ -292,4 +642,299 @@ actual class StringBuilder private constructor (
...
+ * @throws IndexOutOfBoundsException or [IllegalArgumentException] when [startIndex] or [endIndex] is out of range of the [value] character sequence indices or when `startIndex > endIndex`.
+ * @throws IndexOutOfBoundsException if [index] is less than zero or greater than the length of this string builder.
+ */
+@SinceKotlin("1.3")
+@ExperimentalStdlibApi
+@Suppress("EXTENSION_SHADOWED_BY_MEMBER", "NOTHING_TO_INLINE")
+public actual inline fun StringBuilder.insertRange(index: Int, value: CharSequence?, startIndex: Int, endIndex: Int): StringBuilder =
+        this.insertRange(index, value, startIndex, endIndex)
+
+// Method arguments renamings
```
Comment (ilya-g) [inline runtime/src/main/kotlin/kotlin/text/StringBuilder.kt]:
```suggestion
// Method parameters renamings
```
Also they should be deprecated with replacement.

**Pull request merged**